onnxruntime/core/providers/coreml/builders/impl/reduction_op_builder.cc

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Shukant Pal.
+// Licensed under the MIT License.
+
+#include "core/providers/common.h"
+#include "core/providers/shared/utils/utils.h"
+
+#ifdef __APPLE__
+#include "core/providers/coreml/builders/model_builder.h"
+#endif
+#include "core/providers/coreml/builders/helper.h"
+#include "core/providers/coreml/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace coreml {
+
+class ReductionOpBuilder : public BaseOpBuilder {
+#ifdef __APPLE__
+ public:
+  void AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const override;
+
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override;
+#endif
+ private:
+  bool IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                         const logging::Logger& logger) const override;
+};
+
+namespace {
+template <typename T>
+void AddReductionParams(T* params, const std::vector<int64_t>& axes, bool keepdims, bool noop_with_empty_axes) {
+  params->set_keepdims(keepdims);
+
+  for (auto& axis : axes)
+    params->add_axes(axis);
+
+  if (axes.size() == 0 && !noop_with_empty_axes)
+    params->set_reduceall(true);
+}
+} // namespace
```

Comment:
nit: this can go in the `#ifdef __APPLE__` too

Reply:
Done!